### PR TITLE
Feature: Safe App access policy fields

### LINKED ISSUE #351: Safe apps: add a field for different access control types
Context: https://github.com/gnosis/safe-react/issues/2672

We need to know if the app restricts access by source (and therefore can't be loaded in the desktop app) to filter such apps.

The field should be an array of accepted origins and NULL if the app doesn't restrict its access.

The field should be
```
access_control: { type: "NO_RESTRICTIONS" }
or
access_control: { type: 'DOMAIN_RESTRICTION', data: [... ] }
```

It should also accept a new query param `host_domain` that accepts an URL and returns apps that can be displayed for this host

- [ ] Implementation
- [ ] Gitbook doc flag update

### PR DESCRIPTION
**Pull request by mikheevm**
This pull request adds a new Client table (only the `url` field for now) and a field `exclusive_clients` to the Safe App model.

It also adds filtering to the view to allow clients to specify the domain URL in the `GET` request and receive apps guaranteed to work on the client on this domain.

Closes #351 

Original issue https://github.com/gnosis/safe-react/issues/2672

Example request/response:
```
GET /api/v1/safe-apps/?client_url=safe.com

[
   {
      "id":0,
      "url":"https://shaw.net/",
      "name":"Reid, Sanchez and Bailey",
      "iconUrl":"https://placekitten.com/71/525",
      "description":"Focused fresh-thinking Graphic Interface",
      "chainIds":[
         956
      ],
      "provider":"None",
      "accessControl":{
         "type":"DOMAIN_ALLOWLIST",
         "value":[
            "safe.com"
         ]
      }
   },
   {
      "id":1,
      "url":"http://www.hernandez-gonzalez.com/",
      "name":"Haynes LLC",
      "iconUrl":"https://dummyimage.com/71x402",
      "description":"Front-line foreground open architecture",
      "chainIds":[
         5737,
         9036
      ],
      "provider":"None",
      "accessControl":{
         "type":"NO_RESTRICTIONS"
      }
   }
]
```

### CODE REVIEW

File: src/safe_apps/tests/test_views.py
Comment (rmeissner) [inline src/safe_apps/tests/test_views.py:303]: Should we add a test entry that has both clients set and should then also be returned?

File: src/safe_apps/serializers.py
Comment (fmrsabino) [inline src/safe_apps/serializers.py:46]: On the API level, we are always returning `value` even though we know it's going to be empty if `NO_RESTRICTIONS`. If we always end up exposing both `type` and `value` the clients might want to check both even though `NO_RESTRICTIONS` should always have `"value" : []`.

Suggestion: introduce two serialisers (one for each type). You'd check the type on this function and serialize accordingly:

- if `NO_RESTRICTIONS`

```javascript
"accessControl" : {
  "type": "NO_RESTRICTIONS"
}
```

- if `DOMAIN_ALLOWLIST`

```javascript
"accessControl" : {
  "type": "DOMAIN_ALLOWLIST",
  "value" : [ ... ] // domains
}
```
On the client side, they can then associate each type with a different payload. Eg.: if the client gets `DOMAIN_ALLOWLIST` then it should map to a type that has a collection of domains.

File: src/safe_apps/tests/factories.py
Comment (fmrsabino) [inline src/safe_apps/tests/factories.py:42]:
```suggestion
            # A list of clients was passed in, use them
```

File: src/safe_apps/tests/test_views.py
Comment (fmrsabino) [inline src/safe_apps/tests/test_views.py:34]: Instead of using `SafeApp.AccessControlPolicy.NO_RESTRICTIONS`, use the expected `str` on the payload side. This way, we can catch if the API was broken due to a refactor (eg.: if we rename `SafeApp.AccessControlPolicy.NO_RESTRICTIONS` everywhere then the test would be successful even though it might not be the result that we wanted)

File: src/safe_apps/tests/test_views.py
Comment (fmrsabino) [inline src/safe_apps/tests/test_views.py:224]: Why the query parameter here? We don't do any custom handling with it right?

**Review by fmrsabino — CHANGES_REQUESTED** [review]

File: src/safe_apps/tests/test_views.py
Comment (mikheevm) [reply src/safe_apps/tests/test_views.py:224]: It's there to check how the API reacts if you pass an unknown client URL. The `client_url` param is used to return all the apps that work on passed URL

Comment (fmrsabino) [reply src/safe_apps/tests/test_views.py:224]: Sorry you are right. I see that you added it on the view side 👍

File: src/safe_apps/models.py
Comment (fmrsabino) [inline src/safe_apps/models.py:17]: Wondering if `models.URLField` is the best field to be used for the constraints that we have on the query side.

would for example: `http://example.com` be different than `https://example.com`.
Or is it relevant that `https://example.com/path/1` is not the same as `https://example.com`

The restrictions set on the access policy seem to be more on the domain level correct? Going with just a pure domain might also be more clear on the query parameter side: `?client_url=http://example.com` vs `?client_url=https://example.com`?

File: src/safe_apps/tests/test_views.py
Comment (mikheevm) [reply src/safe_apps/tests/test_views.py:303]: I adjusted the existing test to include such app as well

File: src/safe_apps/models.py
Comment (mikheevm) [reply src/safe_apps/models.py:19]: This makes sense to me. Is there a pre-defined domain field for django models? Or I should just go with CharField + custom validator?

Comment (mikheevm) [reply src/safe_apps/models.py:19]: https://code.djangoproject.com/ticket/18120 `wontfix` :/

Comment (fmrsabino) [reply src/safe_apps/models.py:19]: I would say a `CharField` with a custom validator should be enough. If we see that we use the same pattern in more places we can think on creating a custom `Field`.

File: src/safe_apps/models.py
Comment (rmeissner) [inline src/safe_apps/models.py:44]: Also without a protocol?

Comment (mikheevm) [reply src/safe_apps/models.py:44]: the protocol is considered valid now

File: src/safe_apps/models.py
Comment (fmrsabino) [inline src/safe_apps/models.py:11]: This is licensed under GPL so it cannot be redistributed under an MIT license.

Isn't it possible to have simpler regex for this? I see that port numbers are also covered. Do we need those?

Comment (mikheevm) [reply src/safe_apps/models.py]: > This is licensed under GPL so it cannot be redistributed under an MIT license.

Hmm, the code itself is a slightly modified URLValidator from Django with dropped additional checks 🤷‍♂️ 

> Isn't it possible to have simpler regex for this? I see that port numbers are also covered. Do we need those?

I believe it is possible, but from studying existing validators, I understood that it's not as simple as it seems, and I decided not to reinvent the wheel because it's guaranteed I'll miss some edge-case. Anyway, I've just pushed a simpler regex that seems to work for our problem.

File: src/safe_apps/models.py
Comment (fmrsabino) [inline src/safe_apps/models.py:7]: Since this is a constant we should follow PEP8 conventions for the naming https://www.python.org/dev/peps/pep-0008/#constants

Suggestion: `_HOSTNAME_VALIDATOR`

File: src/safe_apps/migrations/0007_safeapp_access_policy.py
Comment (fmrsabino) [inline src/safe_apps/migrations/0007_safeapp_access_policy.py:34]: This regex is different than the one set in the model 🤔 

Maybe it's better to recreate this migration:

```shell
python src/manage.py migrate safe_apps 0006
rm src/safe_apps/migrations/0007_safeapp_access_policy.py
python src/manage.py makemigrations
```

**Review by fmrsabino — CHANGES_REQUESTED** [review]
I think the new regex looks much better! 🚀 

Left some questions/suggestions

File: src/safe_apps/models.py
Comment (mikheevm) [reply src/safe_apps/models.py:7]: I'm curious how I got through that Linter Castle...

File: src/safe_apps/migrations/0007_safeapp_access_policy.py
Comment (mikheevm) [reply src/safe_apps/migrations/0007_safeapp_access_policy.py:34]: It's the same, isn't it? I recreated the migration before pushing and did it once again but the output is the same

Comment (mikheevm) [reply src/safe_apps/migrations/0007_safeapp_access_policy.py:34]: https://github.com/gnosis/safe-config-service/pull/363/commits/3be622b83cd22b5f90557229b34f888774426e11

**Review by fmrsabino — COMMENTED** [review]
Looks good 👏 

Just one question. Did you try adding a `Client` via the admin interface (does it work as expected with the many-to-many relationship)?

**Comment by mikheevm** [comment]
@fmrsabino good catch, I actually forgot to register it to the admin panel. I fixed and it and tested it
Screenshot 2021-12-31 at 17 19 49
 - works as expected
Screenshot 2021-12-31 at 17 19 42

**Review by fmrsabino — APPROVED** [review]
🚀 🚀 🚀

**Comment by mikheevm** [comment]
Thanks for the thorough review, @fmrsabino. Is it good to go now, or should we wait for a review from Uxio?

**Comment by fmrsabino** [comment]
> Thanks for the thorough review, @fmrsabino. Is it good to go now, or should we wait for a review from Uxio?

Let's merge it 🙌 We can start testing with the clients on staging!

**Pull request merged**
